### PR TITLE
All tests pass

### PR DESCRIPTION
**Pull request by ArmchairGraduate**

### CODE REVIEW

File: tests/unit_tests/test_wave_01.py [hunk 1]
```diff
@@ -49,7 +49,13 @@ def test_removing_not_found_is_false():
 
     result = vendor.remove(item)
 
-    raise Exception("Complete this test according to comments below.")
+    # raise Exception
+    # Was I supposed to comment ^^ out?
+    
+    assert len(vendor.inventory) == 3
+    assert item not in vendor.inventory
+    assert result == False
```
Comment (anselrognlie) [inline tests/unit_tests/test_wave_01.py:57]: 👍 
I like the length test to ensure that all the items are still there. We don't really need the check for `item` not being in the list, since it was never in the list. And I like the explicit `False` check here, rather than something a little more falsy.

We could still consider checking that each of the three things that _were_ in the list at the start are still there (that there were no shenanigans from the remove), but given that you're checking the overall length already, that's largely a matter of personal preference.

File: tests/unit_tests/test_wave_02.py [hunk 2]
```diff
@@ -34,7 +34,11 @@ def test_get_no_matching_items_by_category():
 
     items = vendor.get_by_category("electronics")
 
-    raise Exception("Complete this test according to comments below.")
+    #raise Exception("Complete this test according to comments below.")
+
+    
+    assert items == []
```
Comment (anselrognlie) [inline tests/unit_tests/test_wave_02.py:40]: 👍 

This always looks strange to me since I always want to write empty list checks as `not items`, but here in the tests we want to be specific like this. In application Python code, we should use the more permissive checks (`not items`), but in a test, we want to ensure that the thing returned really is a list and that it is empty.

File: tests/unit_tests/test_wave_06.py [hunk 3]
```diff
@@ -76,16 +76,26 @@ def test_swap_best_by_category():
...
+    # raise Exception("Complete this test according to comments below.")
     # *********************************************************************
     # ****** Complete Assert Portion of this test **********
     # *********************************************************************
     # Assertions should check:
     # - That the results is truthy
     # - That tai and jesse's inventories are the correct length
     # - That all the correct items are in tai and jesse's inventories, including the items which were swapped from one vendor to the other
+    
+    assert result == True
```
Comment (anselrognlie) [inline tests/unit_tests/test_wave_06.py:88]: 👍

File: tests/unit_tests/test_wave_06.py [hunk 4]
```diff
@@ -109,16 +119,26 @@ def test_swap_best_by_category_reordered():
...
+    # raise Exception("Complete this test according to comments below.")
     # *********************************************************************
     # ****** Complete Assert Portion of this test **********
     # *********************************************************************
     # Assertions should check:
     # - That result is truthy
     # - That tai and jesse's inventories are the correct length
     # - That all the correct items are in tai and jesse's inventories, and that the items that were swapped are not there
+    
+    assert result is True
```
Comment (anselrognlie) [inline tests/unit_tests/test_wave_06.py:131]: 👍

File: tests/unit_tests/test_wave_06.py [hunk 5]
```diff
@@ -194,16 +214,25 @@ def test_swap_best_by_category_no_match_is_false():
...
-    raise Exception("Complete this test according to comments below.")
+    # raise Exception("Complete this test according to comments below.")
     # *********************************************************************
     # ****** Complete Assert Portion of this test **********
     # *********************************************************************
     # Assertions should check:
     # - That result is falsy
     # - That tai and jesse's inventories are the correct length
     # - That all the correct items are in tai and jesse's inventories
+    assert result is False
```
Comment (anselrognlie) [inline tests/unit_tests/test_wave_06.py:225]: 👍

File: tests/unit_tests/test_wave_06.py [hunk 6]
```diff
@@ -227,11 +256,21 @@ def test_swap_best_by_category_no_other_match_is_false():
...
+    # raise Exception("Complete this test according to comments below.")
     # *********************************************************************
     # ****** Complete Assert Portion of this test **********
     # *********************************************************************
     # Assertions should check:
     # - That result is falsy
     # - That tai and jesse's inventories are the correct length
     # - That all the correct items are in tai and jesse's inventories
+
+    assert result is False
```
Comment (anselrognlie) [inline tests/unit_tests/test_wave_06.py:268]: 👍

File: swap_meet/item.py [hunk 7]
```diff
@@ -1,2 +1,23 @@
+
 class Item:
-    pass
\ No newline at end of file
+    category = ""
+    
+    def __init__(self, category = category, condition = 0):
+        self.category = category
+        self.condition = condition
```
Comment (anselrognlie) [inline swap_meet/item.py:7]: We don't need to use a class variable in the default parameter values here. Strings are immutable, and we can list a string value as a default variable just like we can the integer value for `condition`.

With respect to formatting, keep in mind that [PEP8](https://peps.python.org/pep-0008/#other-recommendations) recommends no spaces around `=` when used for default values or keyword arguments. So we might prefer to write

```python
    def __init__(self, category="", condition=0):
        self.category = category
        self.condition = condition
```

File: swap_meet/item.py [hunk 8]
```diff
@@ -1,2 +1,23 @@
+
 class Item:
-    pass
\ No newline at end of file
+    category = ""
```
Comment (anselrognlie) [inline swap_meet/item.py:3]: Assigning variables in the class like this is actually making a class variable, not an instance variable. We only needed instance variables in this project.

File: swap_meet/item.py [hunk 9]
```diff
@@ -1,2 +1,23 @@
...
+    
+    def __init__(self, category = category, condition = 0):
+        self.category = category
+        self.condition = condition
+
+    def __str__(self):
+        self = "Hello World!"
+        return f"{self}"
```
Comment (anselrognlie) [inline swap_meet/item.py:11]: We shouldn't re-assign `self` within an instance method. Remember that `self` is used to pass in a reference to the instance a method was called on. Re-assigning `self` won't affect that instance, but if we _did_ need to get a value from it, this would make that impossible. Here, we can simply return the desired string directly.

```python
    def __str__(self):
        return "Hello World!"
```

File: swap_meet/item.py [hunk 10]
```diff
@@ -1,2 +1,23 @@
...
+
+    def condition_description(self):
```
Comment (anselrognlie) [inline swap_meet/item.py:13]: This logic will handle anything unexpected below the "normal" range (anything 1 or less will be trash), but what would happen for conditions greater than 5? Maybe the behavior here is OK, but be sure to think about those scenarios.

Also, notice how repetitive all the conditions are. If we wanted to add additional conditions, we would need to write more code to handle them. What if we stored the conditions in a collection and could somehow calculate where in the collection the appropriate description for each condition could be found? If we do it carefully, adding additional conditions could be as simple as adding a new string to the collection, and there might be less code overall that would be easier to test (are ALL of these messages currently being tested?).

File: swap_meet/clothing.py [hunk 11]
```diff
@@ -1,2 +1,14 @@
...
+
+    def __init__(self, condition = 0):
+        super().__init__(condition = condition, category = "Clothing")
+        self.condition = condition
+        
+        
+        
+    def __str__(self):
+        self = "The finest clothing you could wear."
+        return f"{self}"
```
Comment (anselrognlie) [inline swap_meet/clothing.py:13]: As with the parent implementation in `Item`, here, we should _not_ re-assign `self`. We can instead return the desired string directly.

```python
    def __str__(self):
        return "The finest clothing you could wear."
```

File: swap_meet/clothing.py [hunk 12]
```diff
@@ -1,2 +1,14 @@
-class Clothing:
-    pass
\ No newline at end of file
+from swap_meet.item import Item
+
+class Clothing(Item):
+
+    def __init__(self, condition = 0):
+        super().__init__(condition = condition, category = "Clothing")
+        self.condition = condition
```
Comment (anselrognlie) [inline swap_meet/clothing.py:7]: Nice job  removing the category from the constructor, since we don't want to encourage the caller to try to change it. Instead, we pass the literal value that we want to set up to the parent's constructor

Since we're calling the parent's `__init__`, which already uses the `condition` value we pass in to set `self.condition`, we don't need to set it again ourself.

Also following the PEP8 guidelines around keyword and default values, we can rewrite this as

```python
    def __init__(self, condition=0):
        super().__init__(condition=condition, category="Clothing")
```

File: swap_meet/decor.py [hunk 13]
```diff
@@ -1,2 +1,12 @@
-class Decor:
-    pass
\ No newline at end of file
+from swap_meet.item import Item
+
+class Decor(Item):
```
Comment (anselrognlie) [inline swap_meet/decor.py:3]: The feedback provided about `Clothing` also applies here.

File: swap_meet/electronics.py [hunk 14]
```diff
@@ -1,2 +1,11 @@
-class Electronics:
-    pass
+from swap_meet.item import Item
+
+class Electronics(Item):
```
Comment (anselrognlie) [inline swap_meet/electronics.py:3]: The feedback provided about `Clothing` also applies here.

File: swap_meet/vendor.py [hunk 15]
```diff
@@ -1,2 +1,69 @@
+
+from swap_meet.item import Item
+
 class Vendor:
-    pass
\ No newline at end of file
+
+    def __init__(self, inventory=None):
+        if inventory is None:
+            inventory = []
```
Comment (anselrognlie) [inline swap_meet/vendor.py:8]: Remember that we look at this when the integration tests were failing. When the `inventory` parameter has a default value of `[]` (even when referred to through a class variable), _every_ `Vendor` instance created without an explicit inventory was sharing the same list for their inventory. Python uses the same _reference_ of any default value for every function call missing the parameter. So when we need to support making a parameter optional which would normally need a mutable value, instead, we set an immutable sentinel value as the default, then check for that and make a new mutable instance for each call.

File: swap_meet/vendor.py [hunk 16]
```diff
@@ -1,2 +1,69 @@
...
+        self.inventory = inventory
+        
+    def add(self, item):
```
Comment (anselrognlie) [inline swap_meet/vendor.py:11]: 👍

File: swap_meet/vendor.py [hunk 17]
```diff
@@ -1,2 +1,69 @@
...
+        self.inventory.append(item)
+        return item
+
+    def remove(self, item):
+        try:
+            self.inventory.remove(item)
+            return item
+        except Exception:
+            return False
```
Comment (anselrognlie) [inline swap_meet/vendor.py:20]: Great use of error handling to deal with `remove` not finding the item. We should try to make the error we handle as specific as possible in order to avoid silencing unexpected errors. In this case, we can test that the actual error raised by `remove` is `ValueError`, which is what we should use in our `except` clause.

```python
    def remove(self, item):
        try:
            self.inventory.remove(item)
            return item
        except ValueError:
            return False
```

File: swap_meet/vendor.py [hunk 18]
```diff
@@ -1,2 +1,69 @@
...
+
+    def get_by_category(self, category):
+        items_by_category = []
+
+        for item in self.inventory:
+            if item.category == category:
+                items_by_category.append(item)
```
Comment (anselrognlie) [inline swap_meet/vendor.py:27]: Notice this pattern of 
```python
result_list = []
for element in source_list:
    if some_condition(element):
        result_list.append(element)
```

can be rewritten as a list comprehension as
```python
result_list = [element for element in source_list if some_condition(element)]
```

Which here would look like
```python
        items_by_category = [item for item in self.inventory if item.category == category]
```

At first, this may seem more difficult to read, but comprehensions are very common python style, so try getting used to working with them!

File: swap_meet/vendor.py [hunk 19]
```diff
@@ -1,2 +1,69 @@
...
+        
+        return items_by_category
+    
+    def swap_items(self, other_vendor, my_item, their_item):
+        if my_item not in self.inventory or their_item not in other_vendor.inventory:
+            return False
```
Comment (anselrognlie) [inline swap_meet/vendor.py:33]: 👍 

Great use of a guard clause to make sure that both vendors are in a valid configuration for performing the swap.

We _could_ still try to use error handling here, but with the two separate `remove` calls, if the second one failed, we would need to to additional work to restore the condition of the inventories, or take a different approach.

File: swap_meet/vendor.py [hunk 20]
```diff
@@ -1,2 +1,69 @@
...
+        return True
+
+    def swap_first_item(self, other_vendor):
+        if self.inventory == [] or other_vendor.inventory == []:
+            return False
+
+
+        else:
+            self.swap_items(other_vendor, self.inventory[0], other_vendor.inventory[0])
+            return True
```
Comment (anselrognlie) [inline swap_meet/vendor.py:47]: Since the initial check above returns when the needed conditions (non-empty inventories) aren't met, we can treat it as a guard clause, removing the explicit else, and unindenting the following "main" logic.

```python
        self.swap_items(other_vendor, self.inventory[0], other_vendor.inventory[0])
        return True
```

File: swap_meet/vendor.py [hunk 21]
```diff
@@ -1,2 +1,69 @@
...
+
+    
+    def get_best_by_category(self, category):
+        
+        best_condition_so_far = 0
+        
+        list_of_items_by_category = self.get_by_category(category)
```
Comment (anselrognlie) [inline swap_meet/vendor.py:54]: 👍 Nice function reuse to start off with a list of inventory items filtered to the desired category.

File: swap_meet/vendor.py [hunk 22]
```diff
@@ -1,2 +1,69 @@
...
+        if list_of_items_by_category == None:
+            return None
+        for item in list_of_items_by_category:
+            if item.condition > best_condition_so_far:
+                best_condition_so_far = item.condition
+        for item in list_of_items_by_category:
+            if item.condition == best_condition_so_far:
+                return item
```
Comment (anselrognlie) [inline swap_meet/vendor.py:62]: We could do these two loops "at once". We could have a second variable that tracks the best item found so far (initialize it to `None`) and each time we update the condition, also update the item. Then at the end of the loop, our second variable would already have the best item and we could return it immediately.

Note that from a big O perspective, what you've written here and what I described are both O(n) with respect to the number of items. Your approach would translate more directly into an approach using `max`.

File: swap_meet/vendor.py [hunk 23]
```diff
@@ -1,2 +1,69 @@
...
+
+    def swap_best_by_category(self, other, my_priority, their_priority):
+
+        self_best_item = self.get_best_by_category(their_priority)
+        their_best_item = other.get_best_by_category(my_priority)
+
+        return self.swap_items(other, self_best_item, their_best_item)
```
Comment (anselrognlie) [inline swap_meet/vendor.py:69]: 👍 

We can reuse several of our existing functions to handle this very cleanly. Remember that `get_best_by_category` can return `None`, so be sure to convince yourself why `swap_items` will work as we would like, even if we pass `None` in for one or both of the arguments.

**Review by anselrognlie — COMMENTED** [review]
Nice job!

All required waves are complete and all tests are passing (including the tests which need completion)! Your code is well organized, you are using good, descriptive variable names, and you had good code reuse.

There are a few places where it looks like you could review some OOP concepts a bit (the use of `self` in instance methods, class variables which we didn't particularly need in this project, and what is really happening when we call a parent's `__init__`).

Otherwise, the main things I would recommend focusing on are some of the smaller details. Watch for places where your code could be a little more pythonic (guard clauses to reduce nesting, list comprehensions). Also, try looking for patterns of repeated code, and even though a helper function might not be a good fit, think about whether putting the repeated data into a collection of some kind might help us write code that's more DRY (and data-driven!).

Overall, well done!
